### PR TITLE
Add SliverGrid and ScrollGrid

### PR DESCRIPTION
**Pull request by abarth**
This patch adds grid supports to slivers and introduces a ScrollGrid
convenience class for making the common types of scrollable grids.

This patch also deploys ScrollGrid in an example in the Flutter Gallery.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import 'box.dart';
+import 'object.dart';
+import 'sliver.dart';
+import 'sliver_multi_box_adaptor.dart';
+
+class SliverGridGeometry {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_grid.dart]: please document the members on this class already

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+    final int firstIndex = _gridDelegate.getMinChildIndexForScrollOffset(
+        constraints, scrollOffset);
+    final int targetLastIndex = _gridDelegate.getMaxChildIndexForScrollOffset(
+        constraints, targetEndScrollOffset);
+
+    if (firstChild != null) {
+      final int oldFirstIndex = indexOf(firstChild);
+      final int oldLastIndex = indexOf(lastChild);
+      final int leadingGarbage = (firstIndex - oldFirstIndex).clamp(
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_grid.dart]: i'd just put this on one line. I don't think wrapping here is helping readability.

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    for (int index = indexOf(trailingChildWithLayout) + 1; index <
+        targetLastIndex; ++index) {
+      final SliverGridGeometry gridGeometry = _gridDelegate
+          .getGeometryForChildIndex(constraints, index);
+      final BoxConstraints childConstraints = gridGeometry.getBoxConstraints(
+          constraints);
+      RenderBox child = childAfter(trailingChildWithLayout);
+      if (child == null) {
+        child = insertAndLayoutChild(
+            childConstraints, after: trailingChildWithLayout);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_grid.dart]: same here. or put each arg on one line and end with a newline before the )

**Comment by Hixie** [comment]
LGTM, though this stuff is hard to review.

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+class SliverGridGeometry {
+  /// Creates an object that describes the placement of a child in a [RenderSliverGrid].
+  const SliverGridGeometry({
+    this.scrollOffset,
+    this.crossAxisOffset,
+    this.mainAxisExtent,
+    this.crossAxisExtent,
+  });
+
+  /// The scroll offset of the leading edge of the child.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_grid.dart]: Is this relative to the leading edge of the grid sliver (the edge at which scroll offset is zero)?

Comment (abarth) [inline packages/flutter/lib/src/rendering/sliver_grid.dart]: Yes.  Will update.

**Comment by Hixie** [comment]
starting from the commit immediately before this commit, flutter_gallery__build aot_snapshot_build_millis regressed by a substantial amount. Since that test is noisy, this is a potential cause.

**Pull request merged**